### PR TITLE
[PKX-008] Pokemon Evolutions tab

### PR DESCRIPTION
**Pull request by giantiyan**

### CODE REVIEW

File: lib/features/evolution_tab/evolution_tab.dart [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                if ((evolutions?.third_evolution?.length ?? 0) > 0)
+                  Expanded(
+                    child: ListView.builder(
+                      shrinkWrap: true,
+                      itemCount: evolutions?.third_evolution?.length,
+                      itemBuilder: (context, index) => Row(
+                        mainAxisAlignment: MainAxisAlignment.start,
+                        children: [
+                          Padding(
+                            padding: EdgeInsets.all(2),
```
Comment (SaturnoMark) [inline lib/features/evolution_tab/evolution_tab.dart]: const

File: lib/features/evolution_tab/evolution_tab.dart [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            crossAxisAlignment: CrossAxisAlignment.center,
+              children: [
+                Expanded(
+                  child: SingleChildScrollView(
+                    child: Column(
+                      children: [
+                        Image.network(evolutions?.first_evolution?.id
+                                .toString()
+                                .pokemonImage ??
+                            '', ),
```
Comment (SaturnoMark) [inline lib/features/evolution_tab/evolution_tab.dart]: remove the trailing comma after ,

File: lib/api/models/pokemon_evolutions_model.dart [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+part 'pokemon_evolutions_model.freezed.dart';
+
+part 'pokemon_evolutions_model.g.dart';
+
+@freezed
+class PokemonEvolutionsModel with _$PokemonEvolutionsModel {
+  factory PokemonEvolutionsModel({
+    @JsonKey(name: 'first_evolution') PokemonModel? first_evolution,
+    @Default(<PokemonModel>[]) @JsonKey(name: 'second_evolution') List<PokemonModel> second_evolution,
```
Comment (ryle-roasa) [inline lib/api/models/pokemon_evolutions_model.dart]: `List<PokemonModel> secondEvolutions`

File: lib/api/models/pokemon_evolutions_model.dart [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import 'package:pokedex/api/models/pokemon_model.dart';
+
+part 'pokemon_evolutions_model.freezed.dart';
+
+part 'pokemon_evolutions_model.g.dart';
+
+@freezed
+class PokemonEvolutionsModel with _$PokemonEvolutionsModel {
+  factory PokemonEvolutionsModel({
+    @JsonKey(name: 'first_evolution') PokemonModel? first_evolution,
```
Comment (ryle-roasa) [inline lib/api/models/pokemon_evolutions_model.dart]: `PokemonModel? first_evolution`

File: lib/api/models/pokemon_evolutions_model.dart [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    @Default(<PokemonModel>[]) @JsonKey(name: 'second_evolution') List<PokemonModel> second_evolution,
+    @Default(<PokemonModel>[]) @JsonKey(name: 'third_evolution') List<PokemonModel> third_evolution,
```
Comment (ryle-roasa) [inline lib/api/models/pokemon_evolutions_model.dart:13]: thirdEvolutions

File: lib/api/models/pokemon_evolutions_model.dart (hunk 3 above)
Comment (ryle-roasa) [reply lib/api/models/pokemon_evolutions_model.dart]: `@JsonKey(name: 'secondEvolutions')`

File: lib/api/models/pokemon_evolutions_model.dart (hunk 4 above)
Comment (ryle-roasa) [reply lib/api/models/pokemon_evolutions_model.dart]: `@JsonKey(name: 'firstEvolution')`

Comment (ryle-roasa) [reply lib/api/models/pokemon_evolutions_model.dart]: same for below,` secondEvolutions `and `thirdEvolutions`

File: lib/features/evolution_tab/evolution_tab.dart [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  const EvolutionTab({this.evolutions});
+
+  final PokemonEvolutionsModel? evolutions;
+
+  @override
+  Widget build(BuildContext context) {
+    return (evolutions != null)
+        ? Padding(
+            padding: const EdgeInsets.symmetric(horizontal: 18),
+            child: Row(
```
Comment (ryle-roasa) [inline lib/features/evolution_tab/evolution_tab.dart]: Refactor:
```
Row(
   children: [
      firstEvolution,
      if (secondEvolutions.isNotEmpty) Column(secondEvolutions),
      if (thirdEvolutions.isNotEmpty) Column(thirdEvolutions),
   ]
)
```

File: lib/features/evolution_tab/evolution_tab.dart [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return (evolutions != null)
+        ? Padding(
+            padding: const EdgeInsets.symmetric(horizontal: 0),
+            child: Row(
+              crossAxisAlignment: CrossAxisAlignment.center,
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: [
+                SingleChildScrollView(
```
Comment (ryle-roasa) [inline lib/features/evolution_tab/evolution_tab.dart]: Create a separate widget for this

File: lib/features/evolution_tab/widgets/evolution_tile.dart [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  final bool? isNotFirstEvolution;
+
+  @override
+  Widget build(BuildContext context) {
+    return Row(
+      children: [
+        if (isNotFirstEvolution ?? false)
+          (Padding(
+            padding: EdgeInsets.all(2),
+            child: Icon(
```
Comment (ryle-roasa) [inline lib/features/evolution_tab/widgets/evolution_tile.dart]: Is this the same as `Icon(Icons.arrow_forward_ios_rounded)`?

File: lib/features/evolution_tab/widgets/evolution_tile.dart [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+              IconData(
+                0xe09c,
+                fontFamily: 'MaterialIcons',
+                matchTextDirection: true,
+              ),
+            ),
+          )),
+        Column(
+          children: [
+            Image.network(
```
Comment (ryle-roasa) [inline lib/features/evolution_tab/widgets/evolution_tile.dart]: Try using this if its better:

```
CachedNetworkImage(
    imageUrl: '$iconUrl${evolution?.id}.png',
    height: 100,
    width: 100,
    placeholder: (context, url) => const LoadingWidget(),
     errorWidget: (context, url, error) => const Icon(Icons.error),
 )
```

Add `cached_network_image` in pubspec.yaml

File: lib/features/evolution_tab/widgets/evolution_tile.dart [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+              id?.toString().pokemonImage ?? '',
+              height: 100,
+              loadingBuilder: (BuildContext context, Widget child,
+                  ImageChunkEvent? loadingProgress) {
+                if (loadingProgress == null) return child;
+                return Center(child: spinKitRipple);
+              },
+            ),
+            Text(name?.toString().replaceDash.toTitleCase ?? '',
+                style: TextStyle(fontWeight: FontWeight.w500)),
```
Comment (SaturnoMark) [inline lib/features/evolution_tab/widgets/evolution_tile.dart]: trailing comma

File: lib/features/evolution_tab/widgets/evolution_tile.dart (hunk 8 above)
Comment (giantiyan) [reply lib/features/evolution_tab/widgets/evolution_tile.dart]: ah yes same

File: lib/features/evolution_tab/widgets/evolution_column.dart [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+            physics: BouncingScrollPhysics(),
+            child: Column(children: [
+              ...?evolutionsList?.map((pokemon) => EvolutionTile(
+                    name: pokemon.name,
+                    id: pokemon.id,
+                    isNotFirstEvolution: isNotFirstEvolution,
+                  ))
+            ]),
+          )
+        : SingleChildScrollView(
```
Comment (ryle-roasa) [inline lib/features/evolution_tab/widgets/evolution_column.dart]: You can remove this, and the checking `isNotFirstEvolution ?? false` above. Just return lines 16 to 25 (SingleChildScrollView) and set the isNotFirstEvolution to `false`

File: lib/features/evolution_tab/evolution_tab.dart [hunk 12]
```diff
@@ -10,50 +10,25 @@ class EvolutionTab extends StatelessWidget {
...
-                    children: [
-                      EvolutionTile(
-                        name: evolutions?.firstEvolution?.name,
-                        id: evolutions?.firstEvolution?.id,
-                        isNotFirstEvolution: false,
-                      )
-                    ],
-                  ),
+        ? Row(
+            crossAxisAlignment: CrossAxisAlignment.center,
```
Comment (SaturnoMark) [inline lib/features/evolution_tab/evolution_tab.dart]: No need for `CrossAxisAlignment.center,` its a default property for Row

**Review by SaturnoMark — CHANGES_REQUESTED** [review]

**Review by SaturnoMark — APPROVED** [review]

**Pull request merged**